MdeModulePkg/Library/MemoryBucketLib/MemoryBucketLib.c

```diff
@@ -0,0 +1,447 @@
+/**@file
+Library for defining PEI memory buckets. This keeps track of the
+different buckets and the data for the associated HOB.
+For internal use only.
+
+Copyright (c) Microsoft Corporation.
+SPDX-License-Identifier: BSD-2-Clause-Patent
+**/
+
+#include <PiPei.h>
+
+#include <Library/DebugLib.h>
+#include <Library/MemoryBucketLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/HobLib.h>
+
+// Memory types being kept in buckets in PEI.  Currently only runtime types.
+EFI_MEMORY_TYPE  MemoryTypes[4] = {
+  EfiRuntimeServicesCode,
+  EfiRuntimeServicesData,
+  EfiACPIReclaimMemory,
+  EfiACPIMemoryNVS
+};
+
+// PEI memory bucket statistics.  Can be extended if necessary
+EFI_MEMORY_TYPE_STATISTICS  RuntimeMemoryStats[4] = {
+  { 0, MAX_ALLOC_ADDRESS, 0, 0, EfiRuntimeServicesCode, TRUE, TRUE  },   // EfiRuntimeServicesCode
+  { 0, MAX_ALLOC_ADDRESS, 0, 0, EfiRuntimeServicesData, TRUE, TRUE  },   // EfiRuntimeServicesData
+  { 0, MAX_ALLOC_ADDRESS, 0, 0, EfiACPIReclaimMemory,   TRUE, FALSE },   // EfiACPIReclaimMemory
+  { 0, MAX_ALLOC_ADDRESS, 0, 0, EfiACPIMemoryNVS,       TRUE, FALSE }    // EfiACPIMemoryNVS
+};
+
+EFI_PHYSICAL_ADDRESS  CurrentBucketTops[4] = {
+  MAX_ALLOC_ADDRESS, // EfiRuntimeServicesCode
+  MAX_ALLOC_ADDRESS, // EfiRuntimeServicesData
+  MAX_ALLOC_ADDRESS, // EfiACPIReclaimMemory
+  MAX_ALLOC_ADDRESS  // EfiACPIMemoryNVS
+};
+
+// Total number of memory buckets
+UINTN  NumberOfBuckets = 4;

```

Comment:
Is this meant to be a constant? In any case, please use the macro for the base definition, that can be used in statically sizeed structures like the arrays. If this needs to change, initialize it from the macro, otherwise, delete it.